src/sciline/pipeline.py

```diff
@@ -532,72 +531,59 @@ def _set_provider(
                 'Series is a special container reserved for use in conjunction with '
                 'sciline.ParamTable and must not be provided directly.'
             )
-        if (origin := get_origin(key)) is not None:
-            subproviders = self._subproviders.setdefault(origin, {})
-            args = get_args(key)
-            subproviders[args] = provider
-        else:
-            self._providers[key] = provider
+        self._providers[key] = provider
 
     def _get_provider(
         self, tp: Union[Type[T], Item[T]], handler: Optional[ErrorHandler] = None
     ) -> Tuple[Provider, Dict[TypeVar, Key]]:
         handler = handler or HandleAsBuildTimeException()
         explanation: List[str] = []
-        if (provider := self._providers.get(tp)) is not None:
-            return provider, {}
-        elif (origin := get_origin(tp)) is not None and (
-            subproviders := self._subproviders.get(origin)
-        ) is not None:
-            requested = get_args(tp)
-            matches = [
-                (subprovider, bound)
-                for args, subprovider in subproviders.items()
-                if (
-                    bound := _find_bounds_to_make_compatible_type_tuple(requested, args)
-                )
-                is not None
-            ]
-            typevar_counts = [len(bound) for _, bound in matches]
-            min_typevar_count = min(typevar_counts, default=0)
-            matches = [
-                m
-                for count, m in zip(typevar_counts, matches)
-                if count == min_typevar_count
-            ]
-
-            if len(matches) == 1:
-                provider, bound = matches[0]
-                return provider, bound
-            elif len(matches) > 1:
-                matching_providers = [provider.location.name for provider, _ in matches]
-                raise AmbiguousProvider(
-                    f"Multiple providers found for type {tp}."
-                    f" Matching providers are: {matching_providers}."
-                )
-            else:
-                typevars_in_expression = _extract_typevars_from_generic_type(origin)
-                if typevars_in_expression:
-                    explanation = [
-                        ''.join(
-                            map(
-                                str,
+
+        matches = [
+            (provider, bound)
+            for ptype, provider in self._providers.items()
+            if (bound := _find_bounds_to_make_compatible_type_tuple((tp,), (ptype,)))
+            is not None
+        ]
```

Comment:
This replaces a constant-time lookup by a linear-time iteration. And this is called for every requirement, recursively. That is we can the complexity of `Pipeline.get` from something like `O(N)` to `O(N^2)`.

 I do not know how large pipelines will get, but this may be severe problem we should not simply ignore.

Can you somehow keep the rest of your refactor, but still have a constant-time lookup by default, falling back to iteration only if there is no match?

Reply:
You're absolutely correct. We could group the providers by their `origin` type ahead of the iteration to achieve the same effect.

However, I tested this for the esssans workflow using
```python
%%timeit
iofq = pipeline.get(BackgroundSubtractedIofQ)
```
and the result is that the time goes from 2ms to 10ms.

I'd say that's not particularly significant and right now I don't think it is worth complicating the code for.
We can always add this in the future if it becomes an issue.

Reply:
I think I disagree. The ESSsans workflow is "tiny", and I am surprised that even there this takes 5x the time! Since the effect is **quadratic** I do not think we should ignore this.

Why do yousuggest we would have to pre-group? Can't we look for an exact match, simply calling `self._providers.get(tp)` and use that if found?

Reply:
> Can't we look for an exact match, simply calling self._providers.get(tp) and use that if found?

Yeah sure we can do that.

I don't agree 5x is dramatic. Consider that if the effect was truly `O(n^2)` we could expect a slowdown proportional to the number of providers in the pipeline, that is, >50.

Reply:
> > Can't we look for an exact match, simply calling self._providers.get(tp) and use that if found?
> 
> Yeah sure we can do that.
> 
> I don't agree 5x is dramatic. Consider that if the effect was truly `O(n^2)` we could expect a slowdown proportional to the number of providers in the pipeline, that is, >50.

I would not expect 50x, since the code also does *other* things, apart from provider lookup.

But to clarify:

- Switching to an algorithm with much worse scaling (here `O(N^2)` instead of`O(N)` should never be done lightheartedly and without great care, *in particular* if we are not in *full* control over all the contexts where the algorithm is used (see next point).
- Sciline is a *generic* package that is meant to be useful well beyond the context of the neutron-scattering data-reduction workflows that we are writing. Therefore, the benchmark you ran above is not a sufficient data point (or rather, it shows that we should *not* make the change).